### PR TITLE
[sonic-swss] Add support for COPP

### PR DESCRIPTION
**Pull request by hrachya-mellanox**
* Added new class CoppOrch
* Added schema for COPP_TABLE

Signed-off-by: hrachya@mellanox.com

### CODE REVIEW

File: doc/swss-schema.md [hunk 1]
```diff
@@ -319,6 +319,62 @@ and reflects the LAG ports into the redis under: `LAG_TABLE:<team0>:port`
...
+The CPU queue priority is strict priority.
+The policer is created and exclusively owned by the given trap group; it will be not shared (bound to) any other object.
+
+	key = "COPP_TABLE:name"
+	name_list     = name | name,name_list
+	queue         = number; strict queue priority. Higher number means higher priority.
+    packet_action = "drop" | "forward" | "copy" | "copy_cancel" | "trap" | "log" | "deny" | "transit"
+	trap_ids      = name_list; Acceptable values: bgp, lacp, arp, lldp, snmp, ssh, ttl error, ip2me
+    trap_action   = packet_action; trap action which will be applied to all trap_ids.
+
```
Comment (stcheng) [inline doc/swss-schema.md]: remove empty spaces here

File: doc/swss-schema.md [hunk 2]
```diff
@@ -319,6 +319,62 @@ and reflects the LAG ports into the redis under: `LAG_TABLE:<team0>:port`
...
+	
+	;Settings for embedded policer. NOTE - if no policer settings are specified, then no policer is created.
+	meter_type  = "packets" | "bytes"
+	mode		= "sr_tcm" | "tr_tcm" | "storm"
+	color		= "aware" | "blind"
+	cbs         = number ;packets or bytes depending on the meter_type value
+	cir         = number ;packets or bytes depending on the meter_type value
+	pbs         = number ;packets or bytes depending on the meter_type value
+	pir         = number ;packets or bytes depending on the meter_type value
+
```
Comment (stcheng) [inline doc/swss-schema.md]: remove empty spaces here

File: doc/swss-schema.md [hunk 3]
```diff
@@ -319,6 +319,62 @@ and reflects the LAG ports into the redis under: `LAG_TABLE:<team0>:port`
...
+The settings in this table configure trap group, which is assigned a list of trap IDs (protocols), priority of CPU queue priority, and a policer.
+The CPU queue priority is strict priority.
+The policer is created and exclusively owned by the given trap group; it will be not shared (bound to) any other object.
+
+	key = "COPP_TABLE:name"
+	name_list     = name | name,name_list
+	queue         = number; strict queue priority. Higher number means higher priority.
+    packet_action = "drop" | "forward" | "copy" | "copy_cancel" | "trap" | "log" | "deny" | "transit"
+	trap_ids      = name_list; Acceptable values: bgp, lacp, arp, lldp, snmp, ssh, ttl error, ip2me
+    trap_action   = packet_action; trap action which will be applied to all trap_ids.
```
Comment (stcheng) [inline doc/swss-schema.md]: change the above tabs to spaces

File: orchagent/copporch.h [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+const std::string copp_action_value_deny        = "deny";
+const std::string copp_action_value_transit     = "transit";
+
+typedef struct _enum_to_name_t
+{
+    sai_hostif_trap_id_t    trap_id;
+    string                  name;
+} enum_to_name_t;
+
+const vector<enum_to_name_t> valid_trap_id_list = {
```
Comment (stcheng) [inline orchagent/copporch.h]: instead of using vector, using map is better.

File: orchagent/copporch.cpp [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+    return task_process_status::task_success;
+}
+
+void CoppOrch::doTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+    if (consumer.m_toSync.empty()) 
+    {
+        return;
+    }
```
Comment (stcheng) [inline orchagent/copporch.cpp]: remove this part. it is already checked.

File: doc/swss-schema.md (hunk 3 above)
Comment (stcheng) [inline doc/swss-schema.md]: what acceptable values for trap_action? trap?

File: orchagent/copporch.cpp [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+                {
+                    SWSS_LOG_ERROR("Failed to get policer for trap gorup:%s, error:%d\n", key.c_str(), sai_status);
+                    return task_process_status::task_failed;
+                }
+                SWSS_LOG_DEBUG("Obtained policer object:%llx for trap_group:%s", attr.value.oid, key.c_str());
+                policer_id = attr.value.oid;                
+                SWSS_LOG_DEBUG("Applying settings to existing policer:%llx for trap_group:%llx, name:%s", policer_id, m_trap_map[key], key.c_str());
+                if(SAI_NULL_OBJECT_ID == policer_id)
+                {
+                    SWSS_LOG_WARN("Policer settings are specified, however trap group:%llx (name:%s) does not have a policer attached to it.", m_trap_map[key], key.c_str());
```
Comment (stcheng) [inline orchagent/copporch.cpp]: do we want to create a policer here? as the policer settings are specified.

File: orchagent/copporch.cpp [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+    }
+    else if (input == copp_action_value_transit)
+    {
+        action = SAI_PACKET_ACTION_TRANSIT;
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Invalid action specified %s\n", input.c_str());
+        return false;
+    }
```
Comment (lguohan) [inline orchagent/copporch.cpp]: can you use map to do it?

File: orchagent/copporch.cpp [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+    else if (input == copp_policer_mode_tr_tcm)
+    {
+        color = SAI_POLICER_COLOR_SOURCE_BLIND;
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Invalid color specified %s\n", input.c_str());
+        return false;
+    }        
+    return true;
```
Comment (lguohan) [inline orchagent/copporch.cpp]: use map to do it?

File: orchagent/copporch.cpp [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+    }
+    else if (input == copp_policer_mode_storm)
+    {
+        mode = SAI_POLICER_MODE_STORM_CONTROL;
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Invalid mode specified %s\n", input.c_str());
+        return false;
+    }
```
Comment (lguohan) [inline orchagent/copporch.cpp]: use map to simplify this?

File: orchagent/copporch.cpp [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+    else if (input == copp_policer_meter_bytes)
+    {
+        meter_value = SAI_METER_TYPE_BYTES;        
+    }
+    else
+    {
+        SWSS_LOG_DEBUG("invalid meter value :%s\n", input.c_str());
+        return false;
+    }
+    return true;
```
Comment (lguohan) [inline orchagent/copporch.cpp]: use map? std::map<std::string, enum>

you can define a meter map to translate the name to enum.

File: orchagent/copporch.cpp [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+        return false;
+    }
+    if (SAI_STATUS_SUCCESS != (sai_status = sai_policer_api->remove_policer(policer_id)))
+    {
+        SWSS_LOG_ERROR("Failed to remove policer:%llx, error:%d\n", policer_id, sai_status);
+        return false;
+    }
+    return true;
+}
+task_process_status CoppOrch::processCoppRule(Consumer& consumer)
```
Comment (lguohan) [inline orchagent/copporch.cpp]: keep a blank line between functions. there are a few other places to show similiar issue, can you address all of them?

File: orchagent/orch.cpp [hunk 12]
```diff
@@ -117,3 +118,36 @@ void Orch::doTask()
...
+    }
+    istringstream ss(str);
+    string tmp;
+    while (getline(ss, tmp, separator[0]))
+    {
+        SWSS_LOG_DEBUG("extracted token:%s", tmp.c_str());
+        tokens.push_back(tmp);
+    }
+    return true;
+}
```
Comment (lguohan) [inline orchagent/orch.cpp]: there is a tokenize in swss-common, can we use that?

File: orchagent/copporch.cpp [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+            SWSS_LOG_DEBUG("found existing trap_group object:%s", key.c_str());
+            if(!policer_attribs.empty())
+            {
+                attr.id = SAI_HOSTIF_TRAP_GROUP_ATTR_POLICER;
+                attr.value.oid = SAI_NULL_OBJECT_ID;
+                if (SAI_STATUS_SUCCESS != (sai_status = sai_hostif_api->get_trap_group_attribute(m_trap_map[key], 1, &attr)))
+                {
+                    SWSS_LOG_ERROR("Failed to get policer for trap gorup:%s, error:%d\n", key.c_str(), sai_status);
+                    return task_process_status::task_failed;
+                }
```
Comment (lguohan) [inline orchagent/copporch.cpp]: why do you need to retrieve the policer id from the SAI. Whether or not you have created one, your app should know.

File: orchagent/copporch.cpp [hunk 14]
```diff
@@ -0,0 +1,540 @@
...
+                {
+                    SWSS_LOG_ERROR("Failed to apply attribute:%d to trap group:%llx, name:%s, error:%d\n", trap_gr_attr.id, m_trap_map[key], key.c_str(), sai_status);
+                    return task_process_status::task_failed;
+                }
+            }                
+        }
+        else
+        {
+            SWSS_LOG_DEBUG("Creating new trap_group object:%s", key.c_str());
+            sai_object_id_t new_trap = SAI_NULL_OBJECT_ID;
```
Comment (lguohan) [inline orchagent/copporch.cpp]: >SAI_NULL_OBJECT_ID [](start = 39, length = 18)

no need to assign a value here.

File: orchagent/copporch.cpp (hunk 13 above)
Comment (hrachya-mellanox) [inline orchagent/copporch.cpp]: Since the policer is fully owned by trap group, we don't need to keep it separately, we can just query the trap group.

**Comment by stcheng** [comment]
ACK

**Pull request merged**